org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java

```diff
@@ -782,4 +786,194 @@ private String convertDebuggerSourceToClient(String uri) {
+                return new Responses.StackTraceResponseBody(result, 0);
+            }
+            ThreadReference thread = getThread(arguments.threadId);
+            int totalFrames = 0;
+            if (thread != null) {
+                totalFrames = thread.frameCount();
+                if (totalFrames <= arguments.startFrame) {
+                    return new Responses.StackTraceResponseBody(result, totalFrames);
+                }
+                try {
+                    List<StackFrame> stackFrames = arguments.levels == 0
+                            ? thread.frames(arguments.startFrame, totalFrames - arguments.startFrame)
+                            : thread.frames(arguments.startFrame,
+                            Math.min(totalFrames - arguments.startFrame, arguments.levels));
+                    for (int i = 0; i < arguments.levels; i++) {
+                        StackFrame stackFrame = stackFrames.get(arguments.startFrame + i);
+                        int frameId = this.objectPool.addObject(stackFrame.thread(), stackFrame);
+                        Types.StackFrame clientStackFrame = convertDebuggerStackFrameToClient(stackFrame, frameId);
+                        result.add(clientStackFrame);
+                    }
+                } catch (IndexOutOfBoundsException ex) {
+                    // ignore if stack frames overflow
+                    return new Responses.StackTraceResponseBody(result, totalFrames);
+                }
+            }
+            return new Responses.StackTraceResponseBody(result, totalFrames);
+        }
+
+        Responses.ResponseBody scopes(Requests.ScopesArguments arguments) {
+            List<Types.Scope> scopes = new ArrayList<>();
+            StackFrame stackFrame = (StackFrame) this.objectPool.getObjectById(arguments.frameId);
+            if (stackFrame == null) {
+                return new Responses.ScopesResponseBody(scopes);
+            }
+            StackFrameScope localScope = new StackFrameScope(stackFrame, "Local");
+            scopes.add(new Types.Scope(
+                    localScope.getScope(), this.objectPool.addObject(stackFrame.thread(), localScope), false));
+
+            return new Responses.ScopesResponseBody(scopes);
+        }
+
+
+        Responses.ResponseBody variables(Requests.VariablesArguments arguments) throws AbsentInformationException {
+            Map<String, Object> options = new HashMap<>();
+            // TODO: when vscode protocol support customize settings of value format, showQualified should be one of the options.
+            boolean showStaticVariables = true;
+            boolean showQualified = true;
+            if (arguments.format != null && arguments.format.hex) {
+                options.put(NumericFormatter.NUMERIC_FORMAT_OPTION, NumericFormatEnum.HEX);
+            }
+            if (showQualified) {
+                options.put(SimpleTypeFormatter.QUALIFIED_CLASS_NAME_OPTION, showQualified);
+            }
```

Comment:
The code is temporarily here, with the new handler mode, the options should be refactored.